File: npro/npro/report/leads_by_status/leads_by_status.py
# Copyright (c) 2013, GreyCube Technologies and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe


def execute(filters=None):
    return get_columns(filters), get_data(filters)


def get_data(filters):
    return frappe.db.sql(
        """
        select
            status, count(status) count
        from
            tabLead
        {where_conditions}
        group by 
            status""".format(
            where_conditions=get_conditions(filters)
        ),
        filters,
        as_dict=True,
        # debug=True,
    )


def get_conditions(filters):
    conditions = []
    if filters.get("from_date"):
        conditions += ["creation >= %(from_date)s"]
    if filters.get("to_date"):
        conditions += ["creation <= %(to_date)s"]

    return conditions and " where " + " and ".join(conditions) or ""


def get_columns(filters):
    return [
        dict(label="Status", fieldname="status", fieldtype="Data", width=200),
        dict(label="Count", fieldname="count", fieldtype="Int", width=90),
    ]

File: npro/npro/report/leads_by_status/leads_by_status.js
// Copyright (c) 2016, GreyCube Technologies and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Leads By Status"] = {
  filters: [
    {
      fieldname: "from_date",
      label: __("From Date (Lead Creation Date)"),
      fieldtype: "Date",
      default: moment().startOf("month"),
    },
    {
      fieldname: "to_date",
      label: __("To Date (Lead Creation Date)"),
      fieldtype: "Date",
      default: frappe.datetime.get_today(),
    },
    {
      fieldname: "timespan",
      label: __("Lead Creation Date in "),
      fieldtype: "Select",
      options: npro.utils.TIMESPAN_OPTIONS,
      on_change: function (query_report) {
        let date_range = npro.utils.get_date_range(
          query_report.get_values().timespan
        );
        frappe.query_report.set_filter_value({
          from_date: date_range[0],
          to_date: date_range[1],
        });
      },
      default: "This Month",
    },
  ],

  after_datatable_render: function (datatable) {
    const chart_columns = {
      count: "#aee4ff",
    };

    npro.utils.create_chart(
      "status",
      chart_columns,
      datatable,
      frappe.query_report
    );
  },
};
